src/main/java/com/scalar/cassy/transferer/AzureBlobFileUploader.java

```diff
@@ -40,44 +46,74 @@ public AzureBlobFileUploader(BlobContainerAsyncClient blobContainerClient) {
+        .submit(() -> {
+          try {
+            try (InputStream inputStream = new FileInputStream(file.toString())) {
+              blobContainerClient
+                  .getBlobClient(key)
+                  .upload(inputStream, new File(file.toString()).length(), true);
+            }
+          } catch (IOException e) {
+            throw new FileTransferException(e);
+          }
+
+          logger.info("Upload file succeeded : " + file.toString());
+          return null;
+        });
   }
 
   @Override
   public void upload(List<Path> files, BackupConfig config) {
     Path dataDir = Paths.get(config.getDataDir());
-    List<Mono<Void>> toBeUploaded = new ArrayList<>();
+    List<Future> uploads = new ArrayList<>();
     for (Path filePath : files) {
       Path relative = dataDir.relativize(filePath);
       String key = BackupPath.create(config, relative.toString());
       if (requiresUpload(key, filePath)) {
         logger.info("Uploading " + filePath);
-        toBeUploaded.add(
-            blobContainerClient
-                .getBlobAsyncClient(key)
-                .uploadFromFile(filePath.toString(), true)
-                .doOnSuccess(blobProperties -> logger.info("Upload succeeded : " + filePath))
-                .doOnError(
-                    error -> {
-                      throw new FileTransferException("Upload failed : " + filePath, error);
-                    }));
+        uploads.add(
+            executorService
+                .submit(() -> {
+                  try {
+                    try (InputStream inputStream = readStream(filePath)) {
+                      blobContainerClient
+                          .getBlobClient(key)
+                          .upload(inputStream, new File(filePath.toString()).length(),true);
+                    }
+                  } catch (IOException e) {
+                    throw new FileTransferException(e);
+                  }
+                  logger.info("Upload file succeeded : " + filePath.toString());
+                }));
       } else {
         logger.info(filePath + " has been already uploaded.");
       }
 
-      if (toBeUploaded.size() >= ASYNC_FILE_UPLOAD_LIMIT) {
-        // Start uploading files asynchronously and wait for them to complete
-        Mono.when(toBeUploaded).block();
-        toBeUploaded.clear();
+      if (uploads.size() >= ASYNC_FILE_UPLOAD_LIMIT) {
+        uploads.forEach(
+            u -> {
+              try {
+                // Start upload files asynchronously and wait for them to complete
+                u.get();
+              } catch (InterruptedException | ExecutionException e) {
+                throw new FileTransferException(e);
+              }
+            });
+        uploads.clear();
       }
     }
 
-    if (toBeUploaded.size() > 0) {
-      // Start uploading files asynchronously and wait for them to complete
-      Mono.when(toBeUploaded).block();
+    if (uploads.size() > 0) {
+      uploads.forEach(
```

Comment:
This code is pretty redundant with the above code, so it should be refactored.
But, I want to clarify if this newly added code is required first as I asked above.

Reply:
Please fix the redundancy.

Reply:
Sorry for the convenience.
Redundancy issue fixed.